### ISSUE TITLE
Error while trying to cast to Chromecast

### ISSUE THREAD

**Issue by Albibox70** [ejurgensen/forked-daapd#647]
Hi
I cannot stream music from my raspberry pi 2 to any of my chromecast devices.
When trying to select chromecast speakers from Retunes, no sound comes out and the log (in log mode) says: 
forked-daapd[17509]: [  LOG]   player: The Chromecast device 'TV Salotto' FAILED
forked-daapd[17509]: [  LOG]     cast: No heartbeat from 'TV Salotto', shutting down

Attached the log in debug mode.
[chromecast_forked-daapd.log](https://github.com/ejurgensen/forked-daapd/files/2712695/chromecast_forked-daapd.log)

**Comment by Albibox70** [comment]
Hi, did you find a little time to test casting to chromecast in order to help me with my problem?
I'm planning to buy a new Chromecast Audio device to use as a remote speaker, but I must be sure they work togheter, so I would like to test them with my 3 existing Chromecast Video.
Best regards
Alberto

**Comment by ejurgensen** [comment]
I have tested with a regular Chromecast (Video) and with a Google Home (I couldn't locate my Chromecast Audio, maybe I have lent it to someone), and while I couldn't reproduce the issue you have, there are some other things I am not satisfied with. Worst thing being that both devices buffer for a very long time (20-30 secs) before starting playback. I would also like to get rid of the mp3 encoding, which degrades quality and makes it difficult to support sync with other outputs like Airplay. I also want to support 48 kHz like the Chromecast does.

So now I started working on changing the entire way forked-daapd works with Chromecast, and I think that work should also solve your issue. However, it might take a while.

**Comment by Albibox70** [comment]
Ok, no problem to wait for it.
Thanks a lot
Alberto

Il giorno ven 11 gen 2019 alle ore 15:10 ejurgensen <
notifications@github.com> ha scritto:

**Comment by jptrsn** [comment]
Not sure if this will add anything to the conversation, but @ejurgensen I've noticed that all my cast receivers don't automatically start playing when I begin playback, but if I open the cast notification on my phone and press play, the audio stream will start on the device.

**Comment by Albibox70** [comment]
Hi,
once upgraded to the last version 26.5 now forked-daap can reach and stream music without errors to Chromecast devices once selected as remote speakers in Retune.
So I can confirm that streaming to my devices is affected from what you say:
"Worst thing being that both devices buffer for a very long time (20-30 secs) before starting playback .... which degrades quality and makes it difficult to support sync with other outputs like Airplay".
Any news about your work on forked-daap & Chromecast?
Best Regards
Alberto

**Comment by ejurgensen** [comment]
Yes, I have made the changes and I am testing them myself now. With the changes my Chromecast is very quick to start playback, so that is great. The changes are in the player_refactor2 branch. If you would like to help test the changes, I can make you a .deb. I assume you are using Raspbian Stretch?

**Comment by Albibox70** [comment]
Hi, I can try to help you test changes, but... 
I'm a poor linux user so you have to help me to install the new .deb version!
Yes, my system is:

> PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
> NAME="Raspbian GNU/Linux"
> VERSION_ID="9"
> VERSION="9 (stretch)"
> ID=raspbian
> ID_LIKE=debian

If you make a .deb, please send me also the instructions to install it and make it work.
Regards
Alberto

**Comment by ejurgensen** [comment]
Here are instructions for installing the test version, it's basically just downloading and installing:
```
wget http://gyfgafguf.dk/temp/forked-daapd_26.5.80~test1-2_armhf.deb
sudo dpkg -i forked-daapd_26.5.80~test1-2_armhf.deb
```
It will warn you that you are downgrading. I made the version number lower, because if the test version doesn't work for you, you can then easily get back to the release version with "apt update && apt upgrade".

**Comment by Albibox70** [comment]
Ok Thanks  I'll try, test and let you know

Il Gio 28 Mar 2019, 23:32 ejurgensen <notifications@github.com> ha scritto:

**Comment by Albibox70** [comment]
Hi
I've downloaded and installed your .deb package, but once I try to stream to chromecast I reverted to the same error of the previous version (unable to stream to the chromecats device):

RaspDaemon - 6 - forked-daapd[12311]: [  LOG]   player: The Chromecast device 'TV Salotto' FAILED

**Comment by ejurgensen** [comment]
Ok, could you retry with the log level set to debug (in the config file), so I can see the details? Please attach the log file (don't copy/paste the entire thing).

**Comment by Albibox70** [comment]
Hi,
attached a small part of the log file set to debug level.
The device to investigate is "TV Salotto".
Hope this help
Regards
Alberto

Il giorno mar 2 apr 2019 alle ore 12:47 ejurgensen <notifications@github.com>
ha scritto:

**Comment by ejurgensen** [comment]
I don’t see an attachment?

**Comment by Albibox70** [comment]
Hi, 
see attached file (yesterday I made a reply to the email, probably something went wrong with the attachment).
Alberto
[forked-daapd.log](https://github.com/ejurgensen/forked-daapd/files/3042540/forked-daapd.log)

**Comment by ejurgensen** [comment]
Thanks for testing. I think I have spotted the bug and fixed it. Please try with this build instead:
http://gyfgafguf.dk/temp/forked-daapd_26.5.80~test2-2_armhf.deb

**Comment by Albibox70** [comment]
Hi
I've installed and tested: now it casts to Chromecast without hanging but:
- it's out of sync with airport express (it starts playing instantaneously, so it's significantly early);
- with a mix of airplay/chromecast speakers, stopping and plyaing music sometimes fails.
See attached files (may contain more than what you need).
[forked-daapd.log](https://github.com/ejurgensen/forked-daapd/files/3047352/forked-daapd.log)

**Comment by ejurgensen** [comment]
Here's another one, where I think I have fixed the second issue you reported:
http://gyfgafguf.dk/temp/forked-daapd_26.5.80~test3-2_armhf.deb

Sync requires more work and especially investigation. The internet is surprisingly void of information about how to control latency on a Chromecast, but with the new way I am sending audio to it, I think it must be possible.

**Comment by ejurgensen** [comment]
The fix for the original error is now merged into master, so I am closing the issue. Next I will work on Chromecast sync, but it will be a while.
